driver-core/src/main/com/mongodb/MongoCredential.java

```diff
@@ -543,4 +579,101 @@ public String toString() {
                 + ", mechanismProperties=<hidden>"
                 + '}';
     }
+
+    // TODO-OIDC these should take a RequestCallbackContext/RefreshCallbackContext, to allow new fields to be added
+    // TODO-OIDC Duration
+
+    public interface RequestCallback {
+        IdpServerResponse onRequest(@Nullable String principalName, IdpServerInfo serverInfo, int timeoutSeconds);
+    }
+
+    public interface RefreshCallback {
```

Comment:
Are you sure the annotation is suitable? It seems to suggest that whatever it is applied to is not actually thread safe, but here we want to point out that it is thread safe? (At least conditionally.)

There's a caveat here, in that a callback is only thread-safe when used by a single MongoClient. But our motivation for not having a global cache is that we would not expect to use these in multiple clients anyway.

Reply:
> Are you sure the annotation is suitable?

I am, and I provided examples of other people using it on interfaces.

> It seems to suggest that whatever it is applied to is not actually thread safe, but here we want to point out that it is thread safe?

We don't want to point out that callbacks are thread-safe, we want to point out that they don't have to be thread-safe.

> There's a caveat here, in that a callback is only thread-safe when used by a single MongoClient.

The callbacks don't have to be thread-safe when used by a single `MongoClient`. Good point about multiple clients: if a user shares callbacks between multiple clients, then the callbacks must be thread-safe. This means that the thread-safety requirement depends on whether callbacks are shared, and this can't be expressed with annotations, i.e., we should document it in prose.